### ISSUE TITLE
[$250] Workspace - Approve button returns briefly for the admin after approving the expense

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#43783]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.83-1
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/4627303&group_by=cases:section_id&group_id=306206&group_order=asc
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** applause internal team
**Slack conversation:**

## Action Performed:
Precondition: Create a Collect workspace in ND with an expensifail account. Invite a Gmail account as a member. Enable Workflows and add a Plaid Regions bank account.

1. Member: Navigate to https://staging.new.expensify.com/

2. Member: Log in
3. Admin: Log in
4. Member: Submit a manual expense to the workspace chat
5. Admin: Quickly open the IOU and approve it

## Expected Result:
"Pay with Expensify" should be there until the expense is paid.

## Actual Result:
"Pay with Expensify" button appears after approving, after a few seconds, it reverts to "Approve" and back to "Pay with Expensify".

## Workaround:
unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/Expensify/App/assets/43996225/8f7906fb-c568-419d-9c6a-99300e943b83

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01ee0bafa7bb3dc0e3</li>
        <li>Upwork Job ID: 1802955877876368442</li>
        <li>Last Price Increase: 2024-10-22</li>
<li>Automatic offers: </li>
<ul>
<li>ZhenjaHorbach | Reviewer | 102863571</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @techievivek</details>

**Comment by kavimuru** [comment]
@stephanieelliott FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors.

**Comment by kavimuru** [comment]
We think this bug might be related to #wave-collect - Release 1

**Comment by stephanieelliott** [comment]
Waiting for proposals on this one

**Comment by VickyStash** [comment]
Hi, I'm Viktoryia from Callstack - expert contributor group - and I would like to work on this issue.

**Comment by VickyStash** [comment]
I wasn't able to reproduce the issue. _Approve_ button doesn't return for me.

https://github.com/Expensify/App/assets/23176449/f50f7001-37a4-4f17-a929-10ee02586fe3

**Comment by VickyStash** [comment]
Note: I'll be OOO on Monday (24.06) 🌴

**Comment by stephanieelliott** [comment]
Thanks @ZhenjaHorbach! @VickyStash sounds like this is intermittent, maybe try reproing a few times?

**Comment by VickyStash** [comment]
I haven't had a lot of time today for this issue, but I was finally able to reproduce the issue.

It looks like the issue can be caused by a combination of **OpenReport** and **ApproveMoneyRequest** API calls.
So if a user quickly opens the expense quickly and press the _Approve_ button, the following steps can happen:
1. The **OpenReport** API command is called when the expense is opened. In cases when I was able to reproduce this issue this request took **2+ seconds**.
2. During **OpenReport** is processed, user taps on Approve button => local DB got updated with optimistic data for expense Approval. As a result the user see the `Approve` button turns into `Pay`. 
3. The app gets the response of **OpenReport** API call, and the onyx data got updated with the data without approval information since API didn't know about the approval. As a result the user see the `Pay` button turns into `Approve`. 
4. In the background **ApproveMoneyRequest** API call takes around **10-15 seconds**. After that time the app gets the response of the Approval request and the onyx data got updated again with the correct data. As a result the user see the `Approve` button turns into `Pay` again.

I need some time to think about the solution.

![image](https://github.com/Expensify/App/assets/23176449/9b50905b-4c97-4eea-89e4-c2465a0ecb44)
![image](https://github.com/Expensify/App/assets/23176449/f9accf63-6d8e-4e47-b3cb-d2ca0632a504)

**Comment by VickyStash** [comment]
Unfortunately, I haven't had enough time today to investigate this issue more, I'll do my best to take a look tomorrow.

**Comment by stephanieelliott** [comment]
Thanks @VickyStash! Glad we were at least able to nail down how to repro it

**Comment by VickyStash** [comment]
Note: I'll be OOO Jul 1 - Jul 7 🌴

**Comment by stephanieelliott** [comment]
@VickyStash is working on this. Given that it's hard to repro and is mostly a display issue that does not impact app functionality, we'll hold on this while they are OOO next week. Setting this to `Weekly` in the meantime since there will likely be no updates for several days.

**Comment by pasyukevich** [comment]
Hi, I'm Eugene from Callstack - expert contributor group.

I have the capacity and am going to take over the issue

**Comment by stephanieelliott** [comment]
Great, thanks @pasyukevich -- I've assigned you on this issue!

**Comment by stephanieelliott** [comment]
Applying the `New Feature` label to get another BZ member on this while I am OOO til July 10. To catch you up on where we are @sonialiap, expert contributor @pasyukevich just took over this issue and is looking into it.. 

Thanks for watching over this, I'll grab it back from you when I return!

**Comment by pasyukevich** [comment]
I was able to reproduce this behavior and am currently thinking about the solution

**Comment by pasyukevich** [comment]
Today's update: Still working on possible solutions for this problem

**Comment by pasyukevich** [comment]
Today's update:

As already mentioned, we have this issue only in the case of a delayed OpenReport Request during the opening of the expense page.

How this can be fixed:    
  1. Improve response speed on the API side for **OpenReport API** request
  2. Remove optimistic data on the Approve request and add the loading state to the button

Also as I see we have flow with `lastUpdateID` to track orders of updates for Onyx. But this part does not apply to the optimistic data race condition with API update. 

I think that this issue is also connected with this one -[ https://github.com/Expensify/App/issues/37560]( https://github.com/Expensify/App/issues/37560). 

This issue looks global and can be reproducible in other places. 

Should I prepare a proposal with a fix only for this case or focus on a global one?

**Comment by stephanieelliott** [comment]
Yeah @pasyukevich, if that is possible then I think we may as well go for a global fix rather than just fixing this case.

**Comment by pasyukevich** [comment]
Got it, I will switch back to this issue tomorrow

**Comment by stephanieelliott** [comment]
Thanks @pasyukevich!

**Comment by pasyukevich** [comment]
Today's update: 

Work in progress, preparing the proposal

**Comment by pasyukevich** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
When we open the freshly submitted expense from the collect workspace and press the "Approve"  button quickly, we can see that the state of this button will change to "Pay", "Approve" and "Pay"  back.
### What is the root cause of that problem?
When we receive a Response from the "Approve" request straight after the `OpenReport` response we do not apply it and pause the queue as we detect inside the `SaveResponseInOnyx` that the `previousUpdateID` from the response is greater than `lastUpdateIDFromClient`. Because of this we are put on hold update and waiting additionaly for the `getMissingOnyxUpdates` request. In the middle of it, we are applying a response from the `OpenReportRequest` received just a moment before the `ApproveMoneyRequest`  response. Only after the `getMissingOnyxUpdates` request is resolved, we apply the data from the `ApproveMoneyRequest` response.
### What changes do you think we should make in order to solve the problem?
On the API side, we need to have `previousUpdateID` inside the `OpenReport` request. In this case, we will detect the gap between the server and frontend state before showing the "Approve" button. User will be able to send `ApproveMoneyRequestrequest` only if the gap is eliminated
### What alternative solutions did you explore? (Optional)
Change `SaveResponseInOnyx` method and even if we detected the gap between FE state and BE state apply a response for the current request, then put the queue on hold, and run `getMissingOnyxUpdates` request.

**Comment by pasyukevich** [comment]
@stephanieelliott Can we have a review of the proposal from the internal team as @ZhenjaHorbach mentioned?  

One of the proposals relates to the API side

**Comment by stephanieelliott** [comment]
Sure! @techievivek would you mind reviewing this proposal: https://github.com/Expensify/App/issues/43783#issuecomment-2252404214

**Comment by techievivek** [comment]
This one certainly is a tricky one, reviewing the GH to see how we can move forward here, thanks.

**Comment by trjExpensify** [comment]
I'm thinking of closing [this dupe](https://github.com/Expensify/App/issues/45624) in favour of this issue as we've progressed further here. That said, can we confirm this will solve for the `Submit` button on delayed submission, referenced [here](https://github.com/Expensify/App/issues/45624#issuecomment-2276046568). I can't see why not, seems to be the same bug. 

@techievivek please also review [the proposal](https://github.com/Expensify/App/issues/43783#issuecomment-2252404214) so we can move it on. ;)

**Comment by mountiny** [comment]
@danieldoglas @chrispader (I believe Chris worked on part of this) would you be able to check out the proposal too, I am not 100% sure if this is safe to do

> Change SaveResponseInOnyx method and even if we detected the gap between FE state and BE state apply a response for the current request, then put the queue on hold, and run getMissingOnyxUpdates request.

I do not know any specific problem with this but just think there is a reason why its built the way its now

**Comment by techievivek** [comment]
> @techievivek please also review https://github.com/Expensify/App/issues/43783#issuecomment-2252404214 so we can move it on. ;)

I looked through the proposal, but both solutions need verification from experts. Vit has already tagged Daniel, so I will wait for them to review it and share their thoughts.

Also ccing @iwiznia and @tgolen since this change is related to Onyx updates. Thanks.

**Comment by danieldoglas** [comment]
@techievivek I think we need to investigate the method that is being called for approving the report. It's probably not returning all the onyxUpdates it should, which then generates the gap for the next request.

**Comment by tgolen** [comment]
I don't have any immediate thoughts, but I am intently following along!

> > Change SaveResponseInOnyx method and even if we detected the gap between FE state and BE state apply a response for the current request, then put the queue on hold, and run getMissingOnyxUpdates request.

> I do not know any specific problem with this but just think there is a reason why its built the way its now

The reason this was done is so that everything happens to the client in the right order. If there is a gap of Onyx updates, that gap should be applied before the updates in the current response, or else the updates in the gap might overwrite the changes in the current response.

**Comment by iwiznia** [comment]
Tim is right and we should not change that behavior. It seems that in this flow we always go through the `getMissingOnyxUpdates` flow? If so what we need to do is figure out why that happens, as that should be rare and only happen when you missed an onyx update.

**Comment by trjExpensify** [comment]
@techievivek have you been able to investigate this suggestion? 

>@techievivek I think we need to investigate the method that is being called for approving the report. It's probably not returning all the onyxUpdates it should, which then generates the gap for the next request.

**Comment by techievivek** [comment]
No, I haven't been able to triage this again, I will take a look into it by tomorrow.

**Comment by techievivek** [comment]
I looked through the request, and it seems like some updates are missing in the response. As a result, we call `GetMissingOnyxMessages` to fetch those updates.

<img width="826" alt="Screenshot 2024-08-20 at 7 02 29 PM" src="https://github.com/user-attachments/assets/ff0ca734-4519-485b-a4bb-c9ebd2c3d1de">
<img width="1719" alt="Screenshot 2024-08-20 at 7 06 39 PM" src="https://github.com/user-attachments/assets/23b5ab40-c89b-486a-a671-b6a9b45acafd">

**Comment by iwiznia** [comment]
Yeah that sounds wrong, can you debut it further to figure out why that happens?

**Comment by stephanieelliott** [comment]
Hey @techievivek have you had a chance to debug any further on this one?

**Comment by techievivek** [comment]
No updates from last week. I will debug this further this week.

**Comment by trjExpensify** [comment]
Thanks, that would be great because this is pretty janky in a mainline core flow. 👍

**Comment by stephanieelliott** [comment]
Just a note that this issue is linked to an external deadline of Sept 9th, if you can try and prioritize it soon @techievivek!

**Comment by techievivek** [comment]
Hey, I have a few other similar priority issue on my plate, can we look for a volunteer here, thanks. 🙇

**Comment by techievivek** [comment]
I'm currently handling two Fast-API issues: [#419963](https://github.com/Expensify/Expensify/issues/419963) and [#408736](https://github.com/Expensify/Expensify/issues/408736). Additionally, there's a cleanup task related to these issues that needs to be prioritized to fully utilize the improved performance: [#427142](https://github.com/Expensify/Expensify/issues/427142).

**Comment by techievivek** [comment]
There is another wave-control issue on my plate where we are going back and forth with the reviews https://github.com/Expensify/Auth/pull/11996

**Comment by trjExpensify** [comment]
Thanks! 

>I'm currently handling two Fast-API issues: https://github.com/Expensify/Expensify/issues/419963 and https://github.com/Expensify/Expensify/issues/408736.

Both of these are `LOW` and `HIGH` in #fast-apis, which is below `CRITICAL` where this issue is classified in #wave-collect as we focus on bugs in core flows as the deliverable for the SuiteWorld release. 

>There is another wave-control issue on my plate where we are going back and forth with the reviews https://github.com/Expensify/Auth/pull/11996

That one looks like it's attached to the Violations in Auth project which is classified as HIGH in the OP tracker, and #wave-control list of projects in the channel [here](https://expensify.slack.com/archives/C06ML6X0W9L/p1725480340167089).   

So I would suggest this issue takes priority 👍

**Comment by zanyrenney** [comment]
sorry did not mean to touch the label here. readding!

**Comment by trjExpensify** [comment]
@techievivek just wondering if you had any thoughts on my comment here: https://github.com/Expensify/App/issues/43783#issuecomment-2340272990

**Comment by stephanieelliott** [comment]
@techievivek another bump on this -- the two Fast-API issues you referenced were closed yesterday, can we bump the priority of this one now?

**Comment by techievivek** [comment]
Yes, it's now a top priority for me. Since I'm still getting familiar with the flow, it's taking a bit longer, I will update the GH today.

**Comment by techievivek** [comment]
Hmm, I can't reproduce this anymore. I'm going to ask QA if it's still reproducible.

**Comment by techievivek** [comment]
@kavimuru 
> The app has changed a bit, now tester only have the pay button without additional setup, but tester can still repro the issue with the pay button.

I think you will need to be on control policy and have the approval flow set so that when we submit an expense above a certain amount, admins need to manually approve it before paying.

**Comment by techievivek** [comment]
@trjExpensify Are you able to reproduce this? I have been trying to reproduce this since yesterday but have had no luck. Am I holding anything wrong here?

https://github.com/user-attachments/assets/2fb4409a-2205-46f2-bab2-adee4f4c1189

**Comment by techievivek** [comment]
CC @stephanieelliott can you also give it a look and see if you can reproduce it? Thanks.

**Comment by trjExpensify** [comment]
>@techievivek The app has changed a bit, now tester only have the pay button without additional setup, but tester can still repro the issue with the pay button.

QA repro'd 18 hours ago in that video? Let's confirm their setup.

**Comment by trjExpensify** [comment]
@techievivek did you confirm the setup in Slack?

**Comment by stephanieelliott** [comment]
Hey @techievivek, what's the latest here?

**Comment by techievivek** [comment]
I have been ooo for the last four working days. I'm catching up, but updating here that I haven't yet confirmed the setup used by QA on Slack.

**Comment by trjExpensify** [comment]
Okay, well the first diff I see in your video [here](https://github.com/Expensify/App/issues/43783#issuecomment-2355429016) versus the prerequisite steps in the OP (copied below), and QA's video [here](https://github.com/Expensify/App/issues/43783#issuecomment-2353526195), is that it doesn't look like you're testing with a VBBA connected to the workspace. I'd start there. 

>Precondition: Create a Collect workspace in ND with an expensifail account. Invite a Gmail account as a member. Enable Workflows and add a Plaid Regions bank account.

**Comment by techievivek** [comment]
I tested this again following Tom's comment above. While I couldn’t reproduce the original issue mentioned in the GH OP https://github.com/Expensify/App/issues/43783#issue-2353765649, I was able to replicate the problem Applause highlighted here: https://github.com/Expensify/App/issues/43783#issuecomment-2353526195. Specifically, I had to click 'Pay with Expensify' twice for the payment to go through, which is definitely a bug.

**Comment by techievivek** [comment]
https://github.com/user-attachments/assets/065b4bdc-b469-4da9-864c-da388d1358e9

**Comment by trjExpensify** [comment]
Great! It would be ideal to replicate the exact conditions to rule those out for the approve button flash - so don't have dupes to cause `Hold` to be another variable in the mix, and per the OP, the invited member is a gmail account. 

Either way, agree the `Pay` button shouldn't need to be clicked twice. Also, did you keep an eye on the LHN, see how that preview messages goes from approved > paid > approved > paid? 


https://github.com/user-attachments/assets/c2688fd7-ef95-480a-a592-6116f99743f5

**Comment by pasyukevich** [comment]
@techievivek Is there anything that I can help with?

My initial proposal - https://github.com/Expensify/App/issues/43783#issuecomment-2252404214

**Comment by stephanieelliott** [comment]
Hey @techievivek any update? Would it be possible leverage external contributors to help with a fix? @pasyukevich [has mentioned](https://github.com/Expensify/App/issues/43783#issuecomment-2404829426) they are willing to help

**Comment by techievivek** [comment]
This is more of a backend issue, and from my last investigation https://github.com/Expensify/App/issues/43783#issuecomment-2396796522, it seems like we are creating duplicated actions for some reason. I couldn't dive into it further due to fireroom related priorities but I will give it a look today.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by techievivek** [comment]
@trjExpensify I think we can close this and I can rather be assigned to the other GH where the new version of the problem is visible(double tap on pay button).

**Comment by muttmuure** [comment]
Where is the other GH?

**Comment by sakluger** [comment]
@techievivek are you talking about https://github.com/Expensify/App/issues/45624?

**Comment by trjExpensify** [comment]
I think @techievivek is referring to this issue: https://github.com/Expensify/App/issues/47503#issuecomment-2413546487

**Comment by stephanieelliott** [comment]
Hey @techievivek is https://github.com/Expensify/App/issues/47503#issuecomment-2413546487 the one you're referring to? Anyone disagree with consolidating to that issue and closing this one?

**Comment by techievivek** [comment]
The original problem mentioned in this GH issue is no longer reproducible. We're now facing a different issue, which you can find here: https://github.com/Expensify/App/issues/47503#issuecomment-2413546487(same as what you mentioned). I think we can close this issue and I can assign myself to the new one, as I already have some context on it.

**Comment by stephanieelliott** [comment]
Sounds good, thanks @techievivek!
